extensions/blocks/story/edit.native.js

```diff
@@ -91,201 +92,215 @@ class StoryEdit extends React.Component {
 		const updatedMediaFiles = this.replaceNewIdInMediaFilesByOldId(
 			payload.mediaId,
 			payload.mediaServerId,
 			payload.mediaUrl
 		);
-		setAttributes({ mediaFiles: updatedMediaFiles });
-		this.setState({ isUploadInProgress: false });
+		setAttributes( { mediaFiles: updatedMediaFiles } );
+		this.setState( { isUploadInProgress: false } );
 	}
 
-	finishMediaUploadWithFailure(payload) {
-		const { setAttributes } = this.props;
-		// TODO change this
-		// setAttributes( { id: payload.mediaId } );
-		this.setState({ isUploadInProgress: false });
+	finishMediaUploadWithFailure( payload ) {
+		// should anything be done on media upload failure, do it here
+		this.setState( { isUploadInProgress: false, didUploadFail: true } );
 	}
 
 	mediaUploadStateReset() {
-		const { setAttributes } = this.props;
-		// TODO change this
-		// setAttributes( { id: null, src: null } );
-		this.setState({ isUploadInProgress: false });
+		this.setState( { isUploadInProgress: false } );
 	}
 
 	// save state handling methods
-	updateMediaSaveProgress(payload) {
+	updateMediaSaveProgress( payload ) {
 		const { setAttributes } = this.props;
-		if (payload.mediaUrl) {
-			setAttributes({ url: payload.mediaUrl });
+		if ( payload.mediaUrl ) {
+			setAttributes( { url: payload.mediaUrl } );
 		}
-		if (!this.state.isSaveInProgress) {
-			this.setState({ isSaveInProgress: true });
+		if ( ! this.state.isSaveInProgress ) {
+			this.setState( { isSaveInProgress: true } );
 		}
 	}
 
-	replaceMediaUrlInMediaFilesById(mediaId, mediaUrl) {
+	replaceMediaUrlInMediaFilesById( mediaId, mediaUrl ) {
 		const { attributes } = this.props;
-		if (mediaId !== undefined && attributes.mediaFiles !== undefined) {
-			for (let i = 0; i < attributes.mediaFiles.length; i++) {
-				if (attributes.mediaFiles[i].id === mediaId.toString()) {
-					attributes.mediaFiles[i].url = mediaUrl;
-					attributes.mediaFiles[i].link = mediaUrl;
+		const newMediaFiles = this.deepCopyMediaFiles();
+		if ( mediaId !== undefined && newMediaFiles !== undefined ) {
+			for ( let i = 0; i < newMediaFiles.length; i++ ) {
+				if ( newMediaFiles[ i ].id === mediaId.toString() ) {
+					newMediaFiles[ i ].url = mediaUrl;
+					newMediaFiles[ i ].link = mediaUrl;
+					return newMediaFiles;
 				}
 			}
 		}
-		return attributes.mediaFiles;
+		return newMediaFiles;
+	}
+
+	deepCopyMediaFiles() {
+		const { attributes } = this.props;
+		const newMediaFiles = [];
+		if ( attributes.mediaFiles !== undefined ) {
+			for ( let i = 0; i < attributes.mediaFiles.length; i++ ) {
+				// copy to new object
+				newMediaFiles[ i ] = {
+					id: attributes.mediaFiles[ i ].id,
+					url: attributes.mediaFiles[ i ].url,
+					link: attributes.mediaFiles[ i ].link,
+					alt: attributes.mediaFiles[ i ].alt,
+					caption: attributes.mediaFiles[ i ].caption,
+					mime: attributes.mediaFiles[ i ].mime,
+					type: attributes.mediaFiles[ i ].type,
+				};
+			}
+		}
```

Comment:
There's a [cloneDeep](https://lodash.com/docs/#cloneDeep) helper on Lodash that might come handy for this situations.

Reply:
Thank you for suggesting this @etoledom ! I searched the jetpack and gutenberg codebases for other uses of [cloneDeep](https://lodash.com/docs/#cloneDeep) and found no other uses plus [this comment here](https://github.com/WordPress/gutenberg/blob/eed44e8458d81facacb4e3bc4ce2be708651086d/packages/edit-site/src/components/editor/global-styles-renderer.js#L19-L21) which caught my attention. While this wouldn't be much of a problem here given we don't expect to be traversing large arrays or anything like that, I thought it was cleaner to use the `.map()` function since we indeed have to do something with one of those objects (modify some of their properties) while iterating so, seemed better suited for the task at hand 👍 

So I ended up simplifying code a bit by using [`Array.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) as suggested by @guarani  in [this other comment](https://github.com/Automattic/jetpack/pull/17140#discussion_r505738129).

Reply:
Nice! Perfect 🙏